pallene/assignment_conversion.lua

```diff
@@ -0,0 +1,349 @@
+                self:visit_stat(stat)
+            end
+        else
+            -- skip record declarations and type aliases
+            assert(typedecl.match_tag(tl_node._tag, "ast.Toplevel"))
+        end
+    end
+
+    -- add the upvalue box record types to the AST
+    for _, node in ipairs(self.box_records) do
+        table.insert(prog_ast.tls, node)
+    end
+
+    return prog_ast
+end
+
+function Converter:visit_stats(stats)
+    for _, stat in ipairs(stats) do
+        self:visit_stat(stat)
+    end
+end
+
+-- Goes over all the ast.Decls inside the function that have been captured by some nested
+-- function, transforms the decl node itself and all the references made to it.
+function Converter:exit_lambda()
+    local func_info = self.func_stack[#self.func_stack]
+    for _, decl in ipairs(func_info.mutated_decls) do
+        if func_info.captured_decls[decl] then
+            assert(not decl._exported_as)
+
+            -- 1. Create a record type `$T` to hold this captured var.
+            -- 2. Transform the ast.Decl node from `local x = value` to `local x: $T =  { value = value }`
+            -- 3. Go over all the references ever made to this variable and transform them from `ast.Var.Name`
+            --    to ast.Var.Dot
+            local typ = types.T.Record(self:type_name(decl.name), { "value" } , { value = decl._type } )
+            self:add_box_type(decl.loc, typ)
+            decl._type = typ
+
+            local init_exp = self.init_exp_of_decl[decl]
+
+            if init_exp then
+                local value_exp = util.copy_table(init_exp)
+                util.empty_table(init_exp)
+
+                init_exp._tag   = "ast.Exp.InitList"
+                init_exp._type  = typ
+                init_exp.fields = {}
+                init_exp.fields[1] = { name = "value", exp = value_exp }
+            else
+                error("upvalues that are not initialized upon declaration cannot be captured.")
+            end
+
+            for _, ref in ipairs(self.ref_to_decl[decl]) do
+                assert(ref._tag == "ast.Var.Name")
+                local var = util.copy_table(ref)
+                ref._tag = "ast.Var.Dot"
+                ref.exp  =  ast.Exp.Var(ref.loc, var)
+                ref.name = "value"
+
+                ref.exp._type = typ
+            end
+        end
+    end
+
+    assert(#self.func_stack > 1)
+    table.remove(self.func_stack)
+end
+
+function Converter:visit_lambda(lambda)
+    local func_info = FuncInfo()
+    table.insert(self.func_stack, func_info)
+
+    for _, arg in ipairs(lambda.arg_decls) do
+        self:register_decl(arg)
+    end
+
+    self:visit_stats(lambda.body.stats)
+    self:exit_lambda()
+end
+
+function Converter:visit_func(func)
+    assert(func._tag == "ast.FuncStat.FuncStat")
+    local lambda = func.value
+    assert(lambda and lambda._tag == "ast.Exp.Lambda")
+    self:visit_lambda(lambda)
+end
+
+function Converter:convert_exps_of_stat(stat)
```

Comment:
Having a function that only works on some kinds of stats (the ones with an `.exps` field) doesn't smell right. IMO, it's ok to copy paste this for loop in the ast.Stat.Return and the ast.Stat.ForIn, etc.